obspy/clients/filesystem/tsindex.py

```diff
@@ -0,0 +1,1820 @@
+    # create a new Client instance
+    client = Client(db_path, datapath_replace=("^", filepath))
+    t = UTCDateTime("2018-01-01T00:00:00.019500")
+    st = client.get_waveforms("IU", "*", "*", "BHZ", t, t + 1)
+    st.plot()
+
+-------------
+Indexer Usage
+-------------
+
+The :class:`~Indexer` provides a high level
+API for indexing a directory tree of miniSEED files using the IRIS
+`mseedindex <https://github.com/iris-edu/mseedindex/>`_ software.
+
+Initialize a indexer object by supplying the root path to data to be indexed.
+
+>>> from obspy.clients.filesystem.tsindex import Indexer
+>>> from obspy.clients.filesystem.tests.test_tsindex \
+...     import get_test_data_filepath
+>>> # for this example get the file path to test data
+>>> filepath = get_test_data_filepath()
+>>> # create a new Indexer instance
+>>> indexer = Indexer(filepath, filename_pattern='*.mseed')
+
+Index a directory tree of miniSEED files by calling
+:meth:`~Indexer.run`. By default this will
+create a database called ``timeseries.sqlite`` in the current working
+directory. The name of the index database can be changed by supplying the
+``database`` parameter when instantiating the
+:class:`~Indexer` object.
+
+.. code-block:: python
+
+  indexer.run()
+
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+from future.utils import native_str
+
+import os
+from os.path import relpath
+from glob import glob
+import datetime
+import subprocess
+import copyreg
+from multiprocessing import Pool
+import types
+import logging
+import requests
+from collections import namedtuple
+import sqlalchemy as sa
+from sqlalchemy.exc import ResourceClosedError
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.pool import QueuePool
+
+from obspy import UTCDateTime
+from obspy.core.stream import Stream
+from obspy.clients.filesystem.miniseed import _MiniseedDataExtractor, \
+    NoDataError
+from obspy.clients.filesystem.db import _get_tsindex_table, \
+    _get_tsindex_summary_table
+
+
+logger = logging.getLogger('obspy.clients.filesystem.tsindex')
+
+
+def _pickle_method(m):
+    """
+    Allows serializing of class and instance methods.
+    """
+    if m.im_self is None:
+        return getattr, (m.im_class, m.im_func.func_name)
+    else:
+        return getattr, (m.im_self, m.im_func.func_name)
+
+
+copyreg.pickle(types.MethodType, _pickle_method)
+
+
+class Client(object):
+    """
+    Time series extraction client for IRIS tsindex database schema.
+    """
+
+    def __init__(self, database, datapath_replace=None, loglevel="WARNING"):
+        """
+        Initializes the client.
+
+        :type database: str or
+            :class:`~TSIndexDatabaseHandler`
+        :param database: Path to sqlite tsindex database or a
+            TSIndexDatabaseHandler object
+        :type datapath_replace: 2-value tuple(str, str)
```

Comment:
is this standard ?

Reply:
Not too sure about this one, so open to opinions. I'm assuming you're referring to defining the type as `2-value tuple(str, str)`.

Reply:
yes, I don't think "2-value" is a standard way, maybe tuple(str, str) is enough here no ?

Reply:
Changed to tuple(str, str)